### PR TITLE
Fix UI content path without regex

### LINKED ISSUE #9522: ui: ui-content-path flag/config no longer works (1.9.0/.1)
Consul has a configuration flag and config setting called `content-path` in order to change the root directory of where the ui runs from `/ui` to a user configured value (say `/consul`).

As of 1.9.0 the configuration value has no effect on the UI.

On first glance it looks like the value is no longer being passed through to the ui from the binary, therefore although HTTP redirects from say `/` to `/consul` (instead of to `/ui`) work, but the UI doesn't know that the configuration value is set and therefore sends the user back to `/ui/consul` instead.

### PR DESCRIPTION
**Pull request by johncowen**
This PR is an eventual replacement for https://github.com/hashicorp/consul/pull/9523 and fixes #9522.

Following chats with @dnephin, we decided we preferred the overall method to pass variables between binary > ui that we had pre-1.9, which was mostly based on gotemplate rather than using regular expressions.

The pre-1.9 code did have its downsides, mainly as we were using gotemplate to inject the variables, inline, straight into the javascript, and this can cause complications with CSP.

We came up with various alternatives to do away with the dependence on regular expressions, and of all of the options (and combinations of options), this PR was the cleanest and least weird way we came up with.

The approach here avoids the CSP issues by injecting the variables as JSON into a `<script type="application/json">` tag, which avoids any such complications. Then, in a separate externally sourced script, we pick the values we need out of that tag, and then perform the same replacement in Ember's meta tag as we did pre-1.9. Importantly, we only do this with the values we need to inject before ember bootstraps itself, which for the moment is only `rootURL`/`ContentPath`.

As we are moving all our 'binary variables' to the `ui_config` 'concept', I changed the frontend part of the codebase to pick the remaining 'not needed pre ember bootstrap' values from the same `ui_config` JSON, but from within ember - once it has started up. There's potentially a tiny bit more cleanup that can be done on the frontend, but I wanted to get the backend all ironed out and merged in before doing this (this cleanup possibly could be in the distant future). There are also a couple more backend variables we could probably pass in using this approach, but again a piece of future work.

~There are no amends to the go tests here yet as I wanted to get review before looking at that, so tests will fail currently.~

### CODE REVIEW

**Review by mikemorris — COMMENTED** [review]
This does look like a cleaner approach, 👍 

Only comment is that `binaryConfig` feels like a bit of an odd name, would maybe prefer something more generic like `uiConfig`

**Comment by johncowen** [comment]
Hey @mikemorris 

Thanks for taking a look!

> This does look like a cleaner approach, 👍

Yeah I think this is the best we'll get with it.

>Only comment is that binaryConfig feels like a bit of an odd name, would maybe prefer something more generic like uiConfig

😆 you had the exact same thought as I had when I was renaming it, but!

The shape of (what this PR calls) `binaryConfig` looks kinda like this:

```
{
ACLsEnabled: false,
ContentPath: '/ui',
UIConfig: {
  'metrics': 'etc'
}
```

So there are things 'higher up' than `uiConfig`, and I wanted to name it something that didn't make anyone think its the _actual_ `ui_config` as its not - its an object higher up than that.

I'd definitely prefer to move all the settings into the same UIConfig in go/the backend, so:

```
{
UIConfig: {
  ACLsEnabled: false,
  ContentPath: '/ui',
  'metrics': 'etc'
}
```
then I'd definitely rename this property on the frontend to `uiConfig` with things like `uiConfig.ContentPath`. But I think I'd rather that was done at some point after we got this fix in (it's kinda what I was referring to as future work in the PR description)

Happy to do this sooner rather than later if you'd rather tho.

**Comment by johncowen** [comment]
Had another thought on 'naming things'.

How about `operatorConfig`? We already use the `operator` term to mean a similar thing. i.e. config that has been set by the operator (which includes `ui_config`).

That would mean `operatorConfig.UIConfig` instead of `uiConfig.UIConfig` or `binaryConfig.UIConfig`.

LMK and I can change up when I look at the tests for this.

File: agent/uiserver/uiserver_test.go [hunk 1]
```diff
@@ -148,55 +156,55 @@ func TestUIServerIndex(t *testing.T) {
...
-	// so I don't feel too bad but maybe I should.
-	// https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454
-	re := regexp.MustCompile(`<meta name="` + name + `+" content="([^"]*)"`)
-
-	matches := re.FindStringSubmatch(content)
-	require.Len(t, matches, 2, "didn't find the %s meta tag", name)
+	var scriptContent *html.Node
+	var find func(node *html.Node)
+
+	// recur down the tree and pick out <script attrName=ourAttrName>
```
Comment (banks) [inline agent/uiserver/uiserver_test.go:172]:
```suggestion
	// Recurse down the tree and pick out <script attrName=ourAttrName>
```

**Review by banks — COMMENTED** [review]
LGTM. I think `operatorConfig` works better. I saw one nit inline but I think if we change that name this is good.

I also didn't test it with read Consul agent but presumably you did @johncowen ?

**Comment by johncowen** [comment]
Cool, thanks @banks! Yeah, I ran it through on a real agent also, all seems good. Will change up the naming thing shortly

**Comment by johncowen** [comment]
@kaxcode this ended up being more JS than I had expected, would you mind casting you eyes over this also for me?

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by johncowen** [comment]
Manually backported to 1.9.x

**Pull request merged**